sw/airborne/modules/optical_flow/px4flow.c

```diff
@@ -30,64 +31,134 @@
+struct mavlink_msg_req req_flow_rad;
+struct mavlink_msg_req req_heartbeat;
 
 
 // callback function on message reception
 static void decode_optical_flow_msg(struct mavlink_message *msg __attribute__((unused)))
 {
-  optical_flow_available = true;
+  static float quality = 0;
+  static float noise = 0;
+  quality = ((float)optical_flow.quality)/255.0;
+  noise = 1-quality;
+
+  if (quality > PX4FLOW_QUALITY_THRESHOLD) {
+    // flip the axis (if the PX4FLOW is mounted as shown in
+    // https://pixhawk.org/modules/px4flow
+    AbiSendMsgVELOCITY_ESTIMATE(PX4FLOW_VELOCITY_ID,
+                                optical_flow.time_usec,
+                                optical_flow.flow_comp_m_y,
+                                optical_flow.flow_comp_m_x,
+                                0.0f,
+                                noise);
+  }
 
   // positive distance means it's known/valid
   if (optical_flow.ground_distance > 0) {
-    // Y negated to get to the body of the drone
-    AbiSendMsgVELOCITY_ESTIMATE(PX4FLOW_VELOCITY_ID, 0,
-                                (optical_flow.flow_x / optical_flow.ground_distance),
-                                -1.0 * (optical_flow.flow_y / optical_flow.ground_distance),
-                                0.0f,
-                                PX4FLOW_NOISE);
-
     AbiSendMsgAGL(AGL_SONAR_PX4FLOW_ID, optical_flow.ground_distance);
   }
 }
 
-/** Initialization function
+/**
+ * According to https://pixhawk.org/modules/px4flow the PX4flow module outputs
+ * OPTICAL_FLOW_RAD message as well as OPTICAL_FLOW message.
+ * However, on my camera only OPTICAL_FLOW is received. Maybe different firmwares send
+ * different messages?
  */
-void px4flow_init(void)
+static void decode_optical_flow_rad_msg(struct mavlink_message *msg __attribute__((unused)))
 {
-  optical_flow_available = false;
+  // TODO: do something useful with the data here
+  static float timestamp = 0;
+  timestamp = ((float)optical_flow_rad.time_usec) * 0.000001;
+  DOWNLINK_SEND_PX4FLOW_RAD(DefaultChannel, DefaultDevice,
+                        &timestamp,
+                        &optical_flow_rad.sensor_id,
+                        &optical_flow_rad.integration_time_us,
+                        &optical_flow_rad.integrated_x,
+                        &optical_flow_rad.integrated_y,
+                        &optical_flow_rad.integrated_xgyro,
+                        &optical_flow_rad.integrated_ygyro,
+                        &optical_flow_rad.integrated_zgyro,
+                        &optical_flow_rad.temperature,
+                        &optical_flow_rad.quality,
+                        &optical_flow_rad.time_delta_distance_us,
+                        &optical_flow_rad.distance);
+}
+
+static void decode_heartbeat_msg(struct mavlink_message *msg __attribute__((unused)))
+{
+  //TODO: Check whether we have a right version of firmware etc? i.e. heartbeat.autopilot=MAV_AUTOPILOT_PX4
+  DOWNLINK_SEND_HEARTBEAT(DefaultChannel, DefaultDevice,
```

Comment:
I'm not in favor of adding mavlink messages verbatim to the paparazzi messages...